In [1]:
import pandas as pd
import numpy as np
import seaborn as sns

In [2]:
pd.set_option('display.max_rows', 100)

In [3]:
car_df = pd.read_csv('../data/cars.csv')

In [4]:
car_df[car_df.duplicated(subset=['Mileage(miles)'], keep=False)].sort_values(by='Mileage(miles)').head(20)

title  Price  Mileage(miles)  Registration_Year  \
3100         MG MG ZS  18365               1               2023   
3219         MG MG ZS  18515               1               2023   
3021  VOLKSWAGEN POLO  15970           11500               2021   
3143  VOLKSWAGEN POLO  15970           11500               2021   
3167  VOLKSWAGEN POLO  15970           11500               2021   
2606  VOLKSWAGEN POLO  15970           11500               2021   
2844  VOLKSWAGEN POLO  15970           11500               2021   
2395  VOLKSWAGEN POLO  15970           11500               2021   
2325  VOLKSWAGEN POLO  15970           11500               2021   
2324  VOLKSWAGEN POLO  15970           11500               2021   
2312  VOLKSWAGEN POLO  16049           11500               2021   
2954  VOLKSWAGEN POLO  15970           11500               2021   
2789  VOLKSWAGEN POLO  15970           11500               2021   
2646  VOLKSWAGEN POLO  15970           11500               2021   
2751  VOLKSWAGEN POLO  15970           11500               2021   
2833  VOLKSWAGEN POLO  15970           11500               2021   
2595  VOLKSWAGEN POLO  15970           11500               2021   
3321     SUZUKI SWIFT  10707           11589               2020   
3569     SUZUKI SWIFT  10707           11589               2020   
3299     SUZUKI SWIFT  10707           11589               2020   

      Previous Owners      Fuel type  Body type Engine Gearbox  Doors  Seats  \
3100              NaN         Petrol        SUV   1.5L  Manual    5.0    5.0   
3219              NaN         Petrol        SUV   1.5L  Manual    5.0    5.0   
3021              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
3143              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
3167              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2606              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2844              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2395              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2325              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2324              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2312              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2954              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2789              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2646              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2751              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2833              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
2595              1.0         Petrol  Hatchback   1.0L  Manual    5.0    5.0   
3321              1.0  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0   
3569              1.0  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0   
3299              1.0  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0   

     Emission Class Service history  
3100         Euro 6             NaN  
3219         Euro 6             NaN  
3021         Euro 6            Full  
3143         Euro 6            Full  
3167         Euro 6            Full  
2606         Euro 6            Full  
2844         Euro 6            Full  
2395         Euro 6            Full  
2325         Euro 6            Full  
2324         Euro 6            Full  
2312         Euro 6            Full  
2954         Euro 6            Full  
2789         Euro 6            Full  
2646         Euro 6            Full  
2751         Euro 6            Full  
2833         Euro 6            Full  
2595         Euro 6            Full  
3321         Euro 6             NaN  
3569         Euro 6             NaN  
3299         Euro 6             NaN

In [5]:
brand_df = pd.read_csv('../data/brand.csv')

In [6]:
car_df

title  Price  Mileage(miles)  Registration_Year  \
0        SKODA FABIA   6900           70189               2016   
1     VAUXHALL CORSA   1495           88585               2008   
2        HYUNDAI I30    949          137000               2011   
3         MINI HATCH   2395           96731               2010   
4     VAUXHALL CORSA   1000           85000               2013   
...              ...    ...             ...                ...   
3682  RENAULT MEGANE   1395           76202               2006   
3683         AUDI A4   6990          119000               2012   
3684    BMW 3 SERIES   3995          139000               2013   
3685    HONDA ACCORD   1390          179190               2007   
3686  VAUXHALL CORSA   2000           82160               2013   

      Previous Owners Fuel type  Body type Engine Gearbox  Doors  Seats  \
0                 3.0    Diesel  Hatchback   1.4L  Manual    5.0    5.0   
1                 4.0    Petrol  Hatchback   1.2L  Manual    3.0    5.0   
2                 NaN    Petrol  Hatchback   1.4L  Manual    5.0    5.0   
3                 5.0    Petrol  Hatchback   1.4L  Manual    3.0    4.0   
4                 NaN    Diesel  Hatchback   1.3L  Manual    5.0    5.0   
...               ...       ...        ...    ...     ...    ...    ...   
3682              4.0    Petrol  Hatchback   1.6L  Manual    5.0    5.0   
3683              NaN    Petrol     Saloon   2.0L  Manual    4.0    5.0   
3684              NaN    Diesel     Saloon   2.0L  Manual    4.0    5.0   
3685              NaN    Diesel     Estate   2.2L  Manual    5.0    5.0   
3686              7.0    Petrol  Hatchback   1.2L  Manual    5.0    5.0   

     Emission Class Service history  
0            Euro 6             NaN  
1            Euro 4            Full  
2            Euro 5             NaN  
3            Euro 4            Full  
4            Euro 5             NaN  
...             ...             ...  
3682         Euro 4             NaN  
3683         Euro 5             NaN  
3684         Euro 5             NaN  
3685         Euro 4            Full  
3686         Euro 5             NaN  

[3687 rows x 13 columns]

### car_df + brand_df

- brand_df의 title에서 브랜드 추출
- car_df에 해당 brand에 맞는 country 추가

- brand_df의 title 대문자로 변환
- car_df의 title을 뛰어쓰기를 기준으로 맨 앞부분만 추출해서 brand_df의 title과 비교

In [7]:
brand_df['title'] = brand_df['title'].str.upper()

In [8]:
brand_df

title         country
0           SKODA  Czech Republic
1        VAUXHALL  United Kingdom
2         HYUNDAI     South Korea
3            MINI  United Kingdom
4            FORD   United States
5           VOLVO          Sweden
6         PEUGEOT          France
7             BMW         Germany
8         CITROEN          France
9   MERCEDES-BENZ         Germany
10          MAZDA           Japan
11           SAAB          Sweden
12     VOLKSWAGEN         Germany
13          HONDA           Japan
14             MG  United Kingdom
15         TOYOTA           Japan
16           SEAT           Spain
17         NISSAN           Japan
18           ALFA           Italy
19        RENAULT          France
20            KIA     South Korea
21         PROTON        Malaysia
22           FIAT           Italy
23           AUDI         Germany
24     MITSUBISHI           Japan
25          LEXUS           Japan
26           LAND  United Kingdom
27      CHEVROLET   United States
28         SUZUKI           Japan
29          DACIA         Romania
30       DAIHATSU           Japan
31           JEEP   United States
32         JAGUAR  United Kingdom
33       CHRYSLER   United States
34          ROVER  United Kingdom
35             DS          France
36         DAEWOO     South Korea
37          DODGE   United States
38        PORSCHE         Germany
39         SUBARU           Japan
40       INFINITI           Japan
41         ABARTH           Italy
42          SMART         Germany
43         MARCOS  United Kingdom
44       MASERATI           Italy
45      SSANGYONG     South Korea
46        LAGONDA  United Kingdom
47          ISUZU           Japan

In [9]:
car_df['title'].str.split().apply(lambda x: x[0])

0          SKODA
1       VAUXHALL
2        HYUNDAI
3           MINI
4       VAUXHALL
          ...   
3682     RENAULT
3683        AUDI
3684         BMW
3685       HONDA
3686    VAUXHALL
Name: title, Length: 3687, dtype: object

In [10]:
car_df['brand'] = car_df['title'].str.split().apply(lambda x: x[0])

In [11]:
# car_df['model'] = car_df['title'].str.split().apply(lambda x: x[0])
car_df['model'] = car_df['title'].str.split().apply(lambda x: ' '.join(x[1:]))

In [12]:
car_df.drop('title', axis=1, inplace=True)

In [13]:
brand_df.rename({'title':'brand'}, axis=1, inplace=True)

In [14]:
brand_df

brand         country
0           SKODA  Czech Republic
1        VAUXHALL  United Kingdom
2         HYUNDAI     South Korea
3            MINI  United Kingdom
4            FORD   United States
5           VOLVO          Sweden
6         PEUGEOT          France
7             BMW         Germany
8         CITROEN          France
9   MERCEDES-BENZ         Germany
10          MAZDA           Japan
11           SAAB          Sweden
12     VOLKSWAGEN         Germany
13          HONDA           Japan
14             MG  United Kingdom
15         TOYOTA           Japan
16           SEAT           Spain
17         NISSAN           Japan
18           ALFA           Italy
19        RENAULT          France
20            KIA     South Korea
21         PROTON        Malaysia
22           FIAT           Italy
23           AUDI         Germany
24     MITSUBISHI           Japan
25          LEXUS           Japan
26           LAND  United Kingdom
27      CHEVROLET   United States
28         SUZUKI           Japan
29          DACIA         Romania
30       DAIHATSU           Japan
31           JEEP   United States
32         JAGUAR  United Kingdom
33       CHRYSLER   United States
34          ROVER  United Kingdom
35             DS          France
36         DAEWOO     South Korea
37          DODGE   United States
38        PORSCHE         Germany
39         SUBARU           Japan
40       INFINITI           Japan
41         ABARTH           Italy
42          SMART         Germany
43         MARCOS  United Kingdom
44       MASERATI           Italy
45      SSANGYONG     South Korea
46        LAGONDA  United Kingdom
47          ISUZU           Japan

- car_df와 brand_df 병합하기 (merge, left join on brand)

In [15]:
pd.merge(car_df, brand_df, how='left', on='brand')

Price  Mileage(miles)  Registration_Year  Previous Owners Fuel type  \
0      6900           70189               2016              3.0    Diesel   
1      1495           88585               2008              4.0    Petrol   
2       949          137000               2011              NaN    Petrol   
3      2395           96731               2010              5.0    Petrol   
4      1000           85000               2013              NaN    Diesel   
...     ...             ...                ...              ...       ...   
3682   1395           76202               2006              4.0    Petrol   
3683   6990          119000               2012              NaN    Petrol   
3684   3995          139000               2013              NaN    Diesel   
3685   1390          179190               2007              NaN    Diesel   
3686   2000           82160               2013              7.0    Petrol   

      Body type Engine Gearbox  Doors  Seats Emission Class Service history  \
0     Hatchback   1.4L  Manual    5.0    5.0         Euro 6             NaN   
1     Hatchback   1.2L  Manual    3.0    5.0         Euro 4            Full   
2     Hatchback   1.4L  Manual    5.0    5.0         Euro 5             NaN   
3     Hatchback   1.4L  Manual    3.0    4.0         Euro 4            Full   
4     Hatchback   1.3L  Manual    5.0    5.0         Euro 5             NaN   
...         ...    ...     ...    ...    ...            ...             ...   
3682  Hatchback   1.6L  Manual    5.0    5.0         Euro 4             NaN   
3683     Saloon   2.0L  Manual    4.0    5.0         Euro 5             NaN   
3684     Saloon   2.0L  Manual    4.0    5.0         Euro 5             NaN   
3685     Estate   2.2L  Manual    5.0    5.0         Euro 4            Full   
3686  Hatchback   1.2L  Manual    5.0    5.0         Euro 5             NaN   

         brand     model         country  
0        SKODA     FABIA  Czech Republic  
1     VAUXHALL     CORSA  United Kingdom  
2      HYUNDAI       I30     South Korea  
3         MINI     HATCH  United Kingdom  
4     VAUXHALL     CORSA  United Kingdom  
...        ...       ...             ...  
3682   RENAULT    MEGANE          France  
3683      AUDI        A4         Germany  
3684       BMW  3 SERIES         Germany  
3685     HONDA    ACCORD           Japan  
3686  VAUXHALL     CORSA  United Kingdom  

[3687 rows x 15 columns]

In [16]:
car_df

Price  Mileage(miles)  Registration_Year  Previous Owners Fuel type  \
0      6900           70189               2016              3.0    Diesel   
1      1495           88585               2008              4.0    Petrol   
2       949          137000               2011              NaN    Petrol   
3      2395           96731               2010              5.0    Petrol   
4      1000           85000               2013              NaN    Diesel   
...     ...             ...                ...              ...       ...   
3682   1395           76202               2006              4.0    Petrol   
3683   6990          119000               2012              NaN    Petrol   
3684   3995          139000               2013              NaN    Diesel   
3685   1390          179190               2007              NaN    Diesel   
3686   2000           82160               2013              7.0    Petrol   

      Body type Engine Gearbox  Doors  Seats Emission Class Service history  \
0     Hatchback   1.4L  Manual    5.0    5.0         Euro 6             NaN   
1     Hatchback   1.2L  Manual    3.0    5.0         Euro 4            Full   
2     Hatchback   1.4L  Manual    5.0    5.0         Euro 5             NaN   
3     Hatchback   1.4L  Manual    3.0    4.0         Euro 4            Full   
4     Hatchback   1.3L  Manual    5.0    5.0         Euro 5             NaN   
...         ...    ...     ...    ...    ...            ...             ...   
3682  Hatchback   1.6L  Manual    5.0    5.0         Euro 4             NaN   
3683     Saloon   2.0L  Manual    4.0    5.0         Euro 5             NaN   
3684     Saloon   2.0L  Manual    4.0    5.0         Euro 5             NaN   
3685     Estate   2.2L  Manual    5.0    5.0         Euro 4            Full   
3686  Hatchback   1.2L  Manual    5.0    5.0         Euro 5             NaN   

         brand     model  
0        SKODA     FABIA  
1     VAUXHALL     CORSA  
2      HYUNDAI       I30  
3         MINI     HATCH  
4     VAUXHALL     CORSA  
...        ...       ...  
3682   RENAULT    MEGANE  
3683      AUDI        A4  
3684       BMW  3 SERIES  
3685     HONDA    ACCORD  
3686  VAUXHALL     CORSA  

[3687 rows x 14 columns]

In [17]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3687 entries, 0 to 3686
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              3687 non-null   int64  
 1   Mileage(miles)     3687 non-null   int64  
 2   Registration_Year  3687 non-null   int64  
 3   Previous Owners    2276 non-null   float64
 4   Fuel type          3687 non-null   object 
 5   Body type          3687 non-null   object 
 6   Engine             3642 non-null   object 
 7   Gearbox            3687 non-null   object 
 8   Doors              3662 non-null   float64
 9   Seats              3652 non-null   float64
 10  Emission Class     3600 non-null   object 
 11  Service history    540 non-null    object 
 12  brand              3687 non-null   object 
 13  model              3687 non-null   object 
dtypes: float64(3), int64(3), object(8)
memory usage: 403.4+ KB


### check NULL

In [18]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3687 entries, 0 to 3686
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              3687 non-null   int64  
 1   Mileage(miles)     3687 non-null   int64  
 2   Registration_Year  3687 non-null   int64  
 3   Previous Owners    2276 non-null   float64
 4   Fuel type          3687 non-null   object 
 5   Body type          3687 non-null   object 
 6   Engine             3642 non-null   object 
 7   Gearbox            3687 non-null   object 
 8   Doors              3662 non-null   float64
 9   Seats              3652 non-null   float64
 10  Emission Class     3600 non-null   object 
 11  Service history    540 non-null    object 
 12  brand              3687 non-null   object 
 13  model              3687 non-null   object 
dtypes: float64(3), int64(3), object(8)
memory usage: 403.4+ KB


- Previous Owners 
- Engine 
- Doors 
- Seats 
- Emission Class 
- Service history Non Null : 540 -> Unknown으로 null 값 채우기

In [19]:
car_df['Service history'] = car_df['Service history'].fillna('Unknown')

In [20]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3687 entries, 0 to 3686
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              3687 non-null   int64  
 1   Mileage(miles)     3687 non-null   int64  
 2   Registration_Year  3687 non-null   int64  
 3   Previous Owners    2276 non-null   float64
 4   Fuel type          3687 non-null   object 
 5   Body type          3687 non-null   object 
 6   Engine             3642 non-null   object 
 7   Gearbox            3687 non-null   object 
 8   Doors              3662 non-null   float64
 9   Seats              3652 non-null   float64
 10  Emission Class     3600 non-null   object 
 11  Service history    3687 non-null   object 
 12  brand              3687 non-null   object 
 13  model              3687 non-null   object 
dtypes: float64(3), int64(3), object(8)
memory usage: 403.4+ KB


- null이 가장 많은 Previous Owners 확인해보기

In [21]:
car_df['Previous Owners']

0       3.0
1       4.0
2       NaN
3       5.0
4       NaN
       ... 
3682    4.0
3683    NaN
3684    NaN
3685    NaN
3686    7.0
Name: Previous Owners, Length: 3687, dtype: float64

In [22]:
car_df[car_df.duplicated(subset=['Mileage(miles)'], keep=False)].sort_values(by='Mileage(miles)').head(20)

Price  Mileage(miles)  Registration_Year  Previous Owners  \
3100  18365               1               2023              NaN   
3219  18515               1               2023              NaN   
3021  15970           11500               2021              1.0   
3143  15970           11500               2021              1.0   
3167  15970           11500               2021              1.0   
2606  15970           11500               2021              1.0   
2844  15970           11500               2021              1.0   
2395  15970           11500               2021              1.0   
2325  15970           11500               2021              1.0   
2324  15970           11500               2021              1.0   
2312  16049           11500               2021              1.0   
2954  15970           11500               2021              1.0   
2789  15970           11500               2021              1.0   
2646  15970           11500               2021              1.0   
2751  15970           11500               2021              1.0   
2833  15970           11500               2021              1.0   
2595  15970           11500               2021              1.0   
3321  10707           11589               2020              1.0   
3569  10707           11589               2020              1.0   
3299  10707           11589               2020              1.0   

          Fuel type  Body type Engine Gearbox  Doors  Seats Emission Class  \
3100         Petrol        SUV   1.5L  Manual    5.0    5.0         Euro 6   
3219         Petrol        SUV   1.5L  Manual    5.0    5.0         Euro 6   
3021         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
3143         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
3167         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2606         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2844         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2395         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2325         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2324         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2312         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2954         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2789         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2646         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2751         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2833         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
2595         Petrol  Hatchback   1.0L  Manual    5.0    5.0         Euro 6   
3321  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0         Euro 6   
3569  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0         Euro 6   
3299  Petrol Hybrid  Hatchback   1.2L  Manual    5.0    5.0         Euro 6   

     Service history       brand  model  
3100         Unknown          MG  MG ZS  
3219         Unknown          MG  MG ZS  
3021            Full  VOLKSWAGEN   POLO  
3143            Full  VOLKSWAGEN   POLO  
3167            Full  VOLKSWAGEN   POLO  
2606            Full  VOLKSWAGEN   POLO  
2844            Full  VOLKSWAGEN   POLO  
2395            Full  VOLKSWAGEN   POLO  
2325            Full  VOLKSWAGEN   POLO  
2324            Full  VOLKSWAGEN   POLO  
2312            Full  VOLKSWAGEN   POLO  
2954            Full  VOLKSWAGEN   POLO  
2789            Full  VOLKSWAGEN   POLO  
2646            Full  VOLKSWAGEN   POLO  
2751            Full  VOLKSWAGEN   POLO  
2833            Full  VOLKSWAGEN   POLO  
2595            Full  VOLKSWAGEN   POLO  
3321         Unknown      SUZUKI  SWIFT  
3569         Unknown      SUZUKI  SWIFT  
3299         Unknown      SUZUKI  SWIFT

In [23]:
car_df['Previous Owners'] = car_df['Previous Owners'].fillna(0).astype('int')

- 만약 브랜드명이랑 모델명이 같을 때 Doors랑 Seats 수 같은 값으로 채우기
- 최빈값 구하기
- 매칭하기

In [24]:
# car_df['door_mode'] = car_df.groupby(['brand', 'model'])['Doors'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

In [25]:
car_df['Doors'] = car_df.groupby(['brand', 'model'])['Doors'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

In [26]:
# car_df['seat_mode'] = car_df.groupby(['brand', 'model'])['Seats'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

In [27]:
car_df['Seats'] = car_df.groupby(['brand', 'model'])['Seats'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

In [28]:
car_df

Price  Mileage(miles)  Registration_Year  Previous Owners Fuel type  \
0      6900           70189               2016                3    Diesel   
1      1495           88585               2008                4    Petrol   
2       949          137000               2011                0    Petrol   
3      2395           96731               2010                5    Petrol   
4      1000           85000               2013                0    Diesel   
...     ...             ...                ...              ...       ...   
3682   1395           76202               2006                4    Petrol   
3683   6990          119000               2012                0    Petrol   
3684   3995          139000               2013                0    Diesel   
3685   1390          179190               2007                0    Diesel   
3686   2000           82160               2013                7    Petrol   

      Body type Engine Gearbox  Doors  Seats Emission Class Service history  \
0     Hatchback   1.4L  Manual    5.0    5.0         Euro 6         Unknown   
1     Hatchback   1.2L  Manual    5.0    5.0         Euro 4            Full   
2     Hatchback   1.4L  Manual    5.0    5.0         Euro 5         Unknown   
3     Hatchback   1.4L  Manual    3.0    4.0         Euro 4            Full   
4     Hatchback   1.3L  Manual    5.0    5.0         Euro 5         Unknown   
...         ...    ...     ...    ...    ...            ...             ...   
3682  Hatchback   1.6L  Manual    5.0    5.0         Euro 4         Unknown   
3683     Saloon   2.0L  Manual    4.0    5.0         Euro 5         Unknown   
3684     Saloon   2.0L  Manual    4.0    5.0         Euro 5         Unknown   
3685     Estate   2.2L  Manual    4.0    5.0         Euro 4            Full   
3686  Hatchback   1.2L  Manual    5.0    5.0         Euro 5         Unknown   

         brand     model  
0        SKODA     FABIA  
1     VAUXHALL     CORSA  
2      HYUNDAI       I30  
3         MINI     HATCH  
4     VAUXHALL     CORSA  
...        ...       ...  
3682   RENAULT    MEGANE  
3683      AUDI        A4  
3684       BMW  3 SERIES  
3685     HONDA    ACCORD  
3686  VAUXHALL     CORSA  

[3687 rows x 14 columns]

In [29]:
car_df = car_df[~car_df[['Doors', 'Seats']].isnull().any(axis=1)]

In [30]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3665 entries, 0 to 3686
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              3665 non-null   int64  
 1   Mileage(miles)     3665 non-null   int64  
 2   Registration_Year  3665 non-null   int64  
 3   Previous Owners    3665 non-null   int64  
 4   Fuel type          3665 non-null   object 
 5   Body type          3665 non-null   object 
 6   Engine             3622 non-null   object 
 7   Gearbox            3665 non-null   object 
 8   Doors              3665 non-null   float64
 9   Seats              3665 non-null   float64
 10  Emission Class     3596 non-null   object 
 11  Service history    3665 non-null   object 
 12  brand              3665 non-null   object 
 13  model              3665 non-null   object 
dtypes: float64(2), int64(4), object(8)
memory usage: 429.5+ KB


- Engine NULL 확인 (모델&브랜드 확인 후 fillna)
- Emission NULL 확인 (모델&브랜드&연도 확인 후 fillna)

In [31]:
car_df.groupby(['brand', 'model'])['Engine'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

0       1.4L
1       1.2L
2       1.4L
3       1.6L
4       1.2L
        ... 
3682    1.6L
3683    1.8L
3684    2.0L
3685    2.2L
3686    1.2L
Name: Engine, Length: 3665, dtype: object

In [32]:
car_df['Engine']=car_df.groupby(['brand', 'model'])['Engine'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

/tmp/ipykernel_31/753080708.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  car_df['Engine']=car_df.groupby(['brand', 'model'])['Engine'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)


- Engine, Emission Class type object -> float

In [33]:
car_df = car_df[~car_df['Engine'].isnull()]

In [34]:
car_df['Emission Class'] = car_df.groupby(['brand', 'model','Registration_Year'])['Emission Class'].transform(lambda x: x.mode().iloc[0] if not x.mode().empty else np.nan)

In [35]:
car_df[car_df['Emission Class'].isnull()].count()

Price                19
Mileage(miles)       19
Registration_Year    19
Previous Owners      19
Fuel type            19
Body type            19
Engine               19
Gearbox              19
Doors                19
Seats                19
Emission Class        0
Service history      19
brand                19
model                19
dtype: int64

In [36]:
car_df = car_df[~car_df['Emission Class'].isnull()]

In [37]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3614 entries, 0 to 3686
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              3614 non-null   int64  
 1   Mileage(miles)     3614 non-null   int64  
 2   Registration_Year  3614 non-null   int64  
 3   Previous Owners    3614 non-null   int64  
 4   Fuel type          3614 non-null   object 
 5   Body type          3614 non-null   object 
 6   Engine             3614 non-null   object 
 7   Gearbox            3614 non-null   object 
 8   Doors              3614 non-null   float64
 9   Seats              3614 non-null   float64
 10  Emission Class     3614 non-null   object 
 11  Service history    3614 non-null   object 
 12  brand              3614 non-null   object 
 13  model              3614 non-null   object 
dtypes: float64(2), int64(4), object(8)
memory usage: 423.5+ KB


In [38]:
car_df['Emission Class'] = car_df['Emission Class'].str.replace('Euro', '').astype('int')

In [39]:
car_df['Engine'] = car_df['Engine'].str.replace('L', '').astype('float')

In [40]:
car_df.reset_index()

index  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0         0   6900           70189               2016                3   
1         1   1495           88585               2008                4   
2         2    949          137000               2011                0   
3         3   2395           96731               2010                5   
4         4   1000           85000               2013                0   
...     ...    ...             ...                ...              ...   
3609   3682   1395           76202               2006                4   
3610   3683   6990          119000               2012                0   
3611   3684   3995          139000               2013                0   
3612   3685   1390          179190               2007                0   
3613   3686   2000           82160               2013                7   

     Fuel type  Body type  Engine Gearbox  Doors  Seats  Emission Class  \
0       Diesel  Hatchback     1.4  Manual    5.0    5.0               6   
1       Petrol  Hatchback     1.2  Manual    5.0    5.0               4   
2       Petrol  Hatchback     1.4  Manual    5.0    5.0               5   
3       Petrol  Hatchback     1.6  Manual    3.0    4.0               5   
4       Diesel  Hatchback     1.2  Manual    5.0    5.0               5   
...        ...        ...     ...     ...    ...    ...             ...   
3609    Petrol  Hatchback     1.6  Manual    5.0    5.0               4   
3610    Petrol     Saloon     1.8  Manual    4.0    5.0               5   
3611    Diesel     Saloon     2.0  Manual    4.0    5.0               5   
3612    Diesel     Estate     2.2  Manual    4.0    5.0               4   
3613    Petrol  Hatchback     1.2  Manual    5.0    5.0               5   

     Service history     brand     model  
0            Unknown     SKODA     FABIA  
1               Full  VAUXHALL     CORSA  
2            Unknown   HYUNDAI       I30  
3               Full      MINI     HATCH  
4            Unknown  VAUXHALL     CORSA  
...              ...       ...       ...  
3609         Unknown   RENAULT    MEGANE  
3610         Unknown      AUDI        A4  
3611         Unknown       BMW  3 SERIES  
3612            Full     HONDA    ACCORD  
3613         Unknown  VAUXHALL     CORSA  

[3614 rows x 15 columns]

### check duplicate

In [41]:
car_df[car_df.duplicated()].count()

Price                800
Mileage(miles)       800
Registration_Year    800
Previous Owners      800
Fuel type            800
Body type            800
Engine               800
Gearbox              800
Doors                800
Seats                800
Emission Class       800
Service history      800
brand                800
model                800
dtype: int64

In [42]:
car_df = car_df.drop_duplicates(keep='first').sort_values(by='Price')

In [43]:
car_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2814 entries, 3245 to 3212
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price              2814 non-null   int64  
 1   Mileage(miles)     2814 non-null   int64  
 2   Registration_Year  2814 non-null   int64  
 3   Previous Owners    2814 non-null   int64  
 4   Fuel type          2814 non-null   object 
 5   Body type          2814 non-null   object 
 6   Engine             2814 non-null   float64
 7   Gearbox            2814 non-null   object 
 8   Doors              2814 non-null   float64
 9   Seats              2814 non-null   float64
 10  Emission Class     2814 non-null   int64  
 11  Service history    2814 non-null   object 
 12  brand              2814 non-null   object 
 13  model              2814 non-null   object 
dtypes: float64(3), int64(5), object(6)
memory usage: 329.8+ KB


In [44]:
car_df.reset_index()

index  Price  Mileage(miles)  Registration_Year  Previous Owners  \
0      3245    400          120842               2010                4   
1      3278    450          115000               2005                0   
2      1238    495           97000               2011                7   
3      3246    499          111875               2007                4   
4      3242    500          138061               2007                5   
...     ...    ...             ...                ...              ...   
2809   3230  25995           36363               2015                0   
2810   3102  25995             222               2023                0   
2811   3224  28240           45155               2019                1   
2812   2936  30000           46000               2017                2   
2813   3212  33900           92000               2019                1   

     Fuel type  Body type  Engine    Gearbox  Doors  Seats  Emission Class  \
0       Diesel  Hatchback     1.2     Manual    3.0    5.0               5   
1       Diesel  Hatchback     1.6     Manual    5.0    5.0               4   
2       Diesel  Hatchback     1.6     Manual    5.0    5.0               5   
3       Petrol  Hatchback     1.0     Manual    5.0    5.0               4   
4       Petrol  Hatchback     1.0     Manual    5.0    5.0               4   
...        ...        ...     ...        ...    ...    ...             ...   
2809    Diesel        SUV     3.0  Automatic    5.0    5.0               5   
2810    Petrol        SUV     1.5     Manual    5.0    5.0               6   
2811    Diesel        SUV     2.0  Automatic    5.0    5.0               6   
2812    Diesel        SUV     3.0  Automatic    5.0    5.0               6   
2813    Diesel        SUV     2.0  Automatic    5.0    5.0               6   

     Service history          brand                    model  
0            Unknown           SEAT                    IBIZA  
1            Unknown       VAUXHALL                    ASTRA  
2            Unknown       VAUXHALL                    ASTRA  
3            Unknown           FORD                   FIESTA  
4            Unknown        RENAULT                     CLIO  
...              ...            ...                      ...  
2809         Unknown           LAND  ROVER RANGE ROVER SPORT  
2810         Unknown           SEAT                    ATECA  
2811         Unknown            BMW                X3 SERIES  
2812         Unknown           LAND  ROVER RANGE ROVER SPORT  
2813         Unknown  MERCEDES-BENZ                GLE CLASS  

[2814 rows x 15 columns]

In [45]:
### make new dataframe(car_sclaer) of car_df

In [46]:
car_scaler = car_df.copy()

### Use Scaler

In [47]:
car_scaler[['Fuel type','Body type','Gearbox','Service history','brand','model']].nunique()

Fuel type            6
Body type            9
Gearbox              2
Service history      2
brand               45
model              421
dtype: int64

- model은 dummi로 만들기엔 값이 너무 많음 car_scaler에서 drop 하기

In [48]:
car_scaler = car_scaler.drop('model', axis=1)

In [49]:
car_scaler['Fuel type'].unique()

array(['Diesel', 'Petrol', 'Petrol Hybrid', 'Petrol Plug-in Hybrid',
       'Electric', 'Diesel Hybrid'], dtype=object)

In [50]:
car_scaler['Fuel type'].replace({'Petrol Hybrid':'P_Hybrid', 'Petrol Plug-in Hybrid':'P_plugin_Hybrid','Diesel Hybrid':'D_Hybrid'}, inplace=True)

In [51]:
car_scaler['Fuel type']

3245    Diesel
3278    Diesel
1238    Diesel
3246    Petrol
3242    Petrol
         ...  
3230    Diesel
3102    Petrol
3224    Diesel
2936    Diesel
3212    Diesel
Name: Fuel type, Length: 2814, dtype: object

In [52]:
car_scaler = pd.get_dummies(car_scaler, columns = ['Fuel type','Body type','Gearbox','Service history','brand'], drop_first=True)

In [53]:
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler

In [54]:
ss=StandardScaler()

In [55]:
ss.fit(car_scaler)

StandardScaler()

In [56]:
ss.transform(car_scaler)

array([[-1.19375096,  0.93174894, -0.31544444, ..., -0.4029015 ,
        -0.27660639, -0.10893225],
       [-1.18164386,  0.78401725, -1.42446577, ...,  2.48199624,
        -0.27660639, -0.10893225],
       [-1.17074746,  0.32883573, -0.09364018, ...,  2.48199624,
        -0.27660639, -0.10893225],
       ...,
       [ 5.5474839 , -0.98221349,  1.68079394, ..., -0.4029015 ,
        -0.27660639, -0.10893225],
       [ 5.97365392, -0.96084524,  1.23718541, ..., -0.4029015 ,
        -0.27660639, -0.10893225],
       [ 6.91800794,  0.20239642,  1.68079394, ..., -0.4029015 ,
        -0.27660639, -0.10893225]])

In [57]:
car_scaler.columns

Index(['Price', 'Mileage(miles)', 'Registration_Year', 'Previous Owners',
       'Engine', 'Doors', 'Seats', 'Emission Class', 'Fuel type_Diesel',
       'Fuel type_Electric', 'Fuel type_P_Hybrid', 'Fuel type_P_plugin_Hybrid',
       'Fuel type_Petrol', 'Body type_Convertible', 'Body type_Coupe',
       'Body type_Estate', 'Body type_Hatchback', 'Body type_MPV',
       'Body type_Pickup', 'Body type_SUV', 'Body type_Saloon',
       'Gearbox_Manual', 'Service history_Unknown', 'brand_ALFA', 'brand_AUDI',
       'brand_BMW', 'brand_CHEVROLET', 'brand_CHRYSLER', 'brand_CITROEN',
       'brand_DACIA', 'brand_DAEWOO', 'brand_DAIHATSU', 'brand_DODGE',
       'brand_DS', 'brand_FIAT', 'brand_FORD', 'brand_HONDA', 'brand_HYUNDAI',
       'brand_INFINITI', 'brand_JAGUAR', 'brand_JEEP', 'brand_KIA',
       'brand_LAND', 'brand_LEXUS', 'brand_MASERATI', 'brand_MAZDA',
       'brand_MERCEDES-BENZ', 'brand_MG', 'brand_MINI', 'brand_MITSUBISHI',
       'brand_NISSAN', 'brand_PEUGEOT', 'brand_PORSCHE

In [58]:
ss_car_scaler = pd.DataFrame(ss.transform(car_scaler), columns=car_scaler.columns)

In [59]:
ss_car_scaler.head()

Price  Mileage(miles)  Registration_Year  Previous Owners    Engine  \
0 -1.193751        0.931749          -0.315444         1.137289 -0.847944   
1 -1.181644        0.784017          -1.424466        -1.022908  0.014481   
2 -1.170747        0.328836          -0.093640         2.757437  0.014481   
3 -1.169779        0.704993          -0.980857         1.137289 -1.279156   
4 -1.169537        1.367181          -0.980857         1.677339 -1.279156   

      Doors     Seats  Emission Class  Fuel type_Diesel  Fuel type_Electric  \
0 -1.776690  0.151312        0.217779          1.348940           -0.018855   
1  0.588867  0.151312       -0.914995          1.348940           -0.018855   
2  0.588867  0.151312        0.217779          1.348940           -0.018855   
3  0.588867  0.151312       -0.914995         -0.741323           -0.018855   
4  0.588867  0.151312       -0.914995         -0.741323           -0.018855   

   ...  brand_SEAT  brand_SKODA  brand_SMART  brand_SSANGYONG  brand_SUBARU  \
0  ...    7.757185     -0.13033    -0.037729        -0.026669     -0.037729   
1  ...   -0.128913     -0.13033    -0.037729        -0.026669     -0.037729   
2  ...   -0.128913     -0.13033    -0.037729        -0.026669     -0.037729   
3  ...   -0.128913     -0.13033    -0.037729        -0.026669     -0.037729   
4  ...   -0.128913     -0.13033    -0.037729        -0.026669     -0.037729   

   brand_SUZUKI  brand_TOYOTA  brand_VAUXHALL  brand_VOLKSWAGEN  brand_VOLVO  
0      -0.11059     -0.191953       -0.402901         -0.276606    -0.108932  
1      -0.11059     -0.191953        2.481996         -0.276606    -0.108932  
2      -0.11059     -0.191953        2.481996         -0.276606    -0.108932  
3      -0.11059     -0.191953       -0.402901         -0.276606    -0.108932  
4      -0.11059     -0.191953       -0.402901         -0.276606    -0.108932  

[5 rows x 67 columns]

In [60]:
ss_car_scaler.describe()

Price  Mileage(miles)  Registration_Year  Previous Owners  \
count  2.814000e+03    2.814000e+03       2.814000e+03      2814.000000   
mean  -1.616018e-16    1.414015e-16      -4.848053e-16         0.000000   
std    1.000178e+00    1.000178e+00       1.000178e+00         1.000178   
min   -1.193751e+00   -2.124062e+00      -4.086117e+00        -1.022908   
25%   -6.975414e-01   -6.091100e-01      -7.590530e-01        -1.022908   
50%   -3.232503e-01   -1.418148e-02      -9.364018e-02         0.057191   
75%    4.031759e-01    5.349697e-01       5.717726e-01         0.597240   
max    6.918008e+00    2.594797e+01       2.568011e+00         3.837536   

             Engine         Doors         Seats  Emission Class  \
count  2.814000e+03  2.814000e+03  2.814000e+03    2.814000e+03   
mean  -1.767519e-16 -1.300389e-16 -4.197858e-16   -4.444048e-16   
std    1.000178e+00  1.000178e+00  1.000178e+00    1.000178e+00   
min   -1.710369e+00 -2.959468e+00 -5.062465e+00   -4.313320e+00   
25%   -8.479438e-01 -5.939111e-01  1.513119e-01   -9.149955e-01   
50%    1.448101e-02  5.888673e-01  1.513119e-01    2.177794e-01   
75%    8.769058e-01  5.888673e-01  1.513119e-01    2.177794e-01   
max    8.423123e+00  5.888673e-01  3.627163e+00    1.350554e+00   

       Fuel type_Diesel  Fuel type_Electric  ...    brand_SEAT   brand_SKODA  \
count      2.814000e+03        2.814000e+03  ...  2.814000e+03  2.814000e+03   
mean      -2.020022e-17       -7.575082e-18  ... -3.787541e-18  3.408787e-17   
std        1.000178e+00        1.000178e+00  ...  1.000178e+00  1.000178e+00   
min       -7.413228e-01       -1.885450e-02  ... -1.289127e-01 -1.303300e-01   
25%       -7.413228e-01       -1.885450e-02  ... -1.289127e-01 -1.303300e-01   
50%       -7.413228e-01       -1.885450e-02  ... -1.289127e-01 -1.303300e-01   
75%        1.348940e+00       -1.885450e-02  ... -1.289127e-01 -1.303300e-01   
max        1.348940e+00        5.303772e+01  ...  7.757185e+00  7.672831e+00   

        brand_SMART  brand_SSANGYONG  brand_SUBARU  brand_SUZUKI  \
count  2.814000e+03     2.814000e+03  2.814000e+03  2.814000e+03   
mean   1.010011e-17     1.262514e-17  1.262514e-17 -2.777530e-17   
std    1.000178e+00     1.000178e+00  1.000178e+00  1.000178e+00   
min   -3.772913e-02    -2.666904e-02 -3.772913e-02 -1.105903e-01   
25%   -3.772913e-02    -2.666904e-02 -3.772913e-02 -1.105903e-01   
50%   -3.772913e-02    -2.666904e-02 -3.772913e-02 -1.105903e-01   
75%   -3.772913e-02    -2.666904e-02 -3.772913e-02 -1.105903e-01   
max    2.650472e+01     3.749667e+01  2.650472e+01  9.042384e+00   

       brand_TOYOTA  brand_VAUXHALL  brand_VOLKSWAGEN   brand_VOLVO  
count  2.814000e+03    2.814000e+03      2.814000e+03  2.814000e+03  
mean  -4.040044e-17   -2.525027e-17      2.525027e-17  8.143214e-17  
std    1.000178e+00    1.000178e+00      1.000178e+00  1.000178e+00  
min   -1.919531e-01   -4.029015e-01     -2.766064e-01 -1.089323e-01  
25%   -1.919531e-01   -4.029015e-01     -2.766064e-01 -1.089323e-01  
50%   -1.919531e-01   -4.029015e-01     -2.766064e-01 -1.089323e-01  
75%   -1.919531e-01   -4.029015e-01     -2.766064e-01 -1.089323e-01  
max    5.209607e+00    2.481996e+00      3.615245e+00  9.180018e+00  

[8 rows x 67 columns]

In [61]:
# ss.fit_transform() 도 있음

In [62]:
from sklearn.decomposition import PCA

In [63]:
pca=PCA()

In [64]:
pca.fit(ss_car_scaler)

PCA()

In [65]:
pd.DataFrame(pca.transform(ss_car_scaler))

0         1         2         3         4         5         6   \
0    -0.266493  0.815977  0.016632 -2.399253 -1.427912  0.514063 -1.696900   
1    -0.138014  0.983745 -1.752224 -1.383354 -0.458305  0.026144  0.102061   
2    -0.163536  0.153468 -1.575657 -1.730911 -0.714790  0.233513  0.104127   
3    -2.152074  1.534897 -1.458435 -0.313508 -0.067436  0.582891 -0.786897   
4    -2.018498  1.595522 -1.651028 -0.253460 -0.057336 -0.561904 -0.097388   
...        ...       ...       ...       ...       ...       ...       ...   
2809  5.063530 -3.818874  0.600426  3.753090 -1.385811  0.935741 -1.216596   
2810  0.180842 -5.270160  2.122769  1.812675  0.307228 -0.067510 -0.573835   
2811  3.994332 -4.528456  1.573923  1.358624  0.323015  0.397683 -0.309974   
2812  5.270724 -4.754567  0.945748  3.558104 -1.363949  1.066410 -1.231484   
2813  4.419921 -4.696911  1.900735  1.140088  0.781772 -0.503691 -0.579953   

            7         8         9   ...        57        58        59  \
0    -0.068923  0.409348 -0.698374  ... -0.377378  0.437198 -0.002521   
1    -0.347526 -0.818380 -0.565909  ... -0.274601 -0.093341  0.040021   
2    -0.717670 -1.342838 -0.747343  ... -0.226355  0.151415  0.084495   
3     0.054380 -0.524099  0.326912  ... -0.266318  0.123246 -0.317276   
4    -0.101299  0.130385  0.804040  ... -0.431820  0.069348 -0.051571   
...        ...       ...       ...  ...       ...       ...       ...   
2809 -0.907259 -1.321868  0.410198  ...  1.579442  0.252419  0.541949   
2810  0.409555  0.422547 -0.465343  ... -1.813439 -1.004037  0.154177   
2811 -0.318151 -1.028498  0.963825  ... -1.516272 -0.621115  0.759155   
2812 -1.064422 -1.535348  0.325916  ...  1.596096  0.415284  0.560953   
2813 -1.216299 -0.564465  0.559802  ... -1.631729 -1.557593  0.717234   

            60        61        62        63        64        65        66  
0    -0.175262 -0.119024 -0.539756  0.237474 -0.002587  0.004254  0.000743  
1     0.047652 -0.054821  0.317798  0.442479  0.001617  0.001689 -0.007505  
2     0.340840  0.149094 -0.262251  0.033279 -0.000839  0.003540 -0.002632  
3    -0.579556 -0.019633  0.039102  0.076248  0.000210  0.001532  0.000252  
4    -0.200195 -0.118487  0.315052  0.173029 -0.001882  0.000782  0.000227  
...        ...       ...       ...       ...       ...       ...       ...  
2809  0.588573 -0.278373  1.793244  0.457293 -0.011574 -0.014975 -0.004133  
2810 -0.140298  0.926648  2.205407  0.089734 -0.004797 -0.015279 -0.002889  
2811 -0.799109 -0.703381  2.147975  0.592412 -0.013001 -0.023176 -0.003609  
2812  0.690213 -0.296219  2.047764  1.006259 -0.016517 -0.018131 -0.002260  
2813 -0.795242 -1.042457  3.375552  1.102331 -0.018029 -0.023028 -0.006099  

[2814 rows x 67 columns]

In [66]:
pca=PCA(5)

In [67]:
pca.fit_transform(car_scaler)

array([[ 3.70408552e+04, -3.14404885e+03, -2.76794198e+00,
         1.65701535e+00,  1.32203990e+00],
       [ 3.12032620e+04, -3.37636014e+03,  2.21966169e+00,
        -2.37578201e+00, -7.62122237e-01],
       [ 1.32221079e+04, -4.20105630e+03, -3.57768429e+00,
         4.77448527e+00, -4.36944751e-01],
       ...,
       [-3.99028135e+04,  2.10067135e+04,  1.07787255e+01,
         6.36290581e-01, -1.05109083e+00],
       [-3.91438324e+04,  2.28054811e+04,  1.41701433e+01,
         1.71061226e+00, -1.23475163e+00],
       [ 6.61402539e+03,  2.89233608e+04,  1.48822925e+01,
         7.31274401e-01, -1.18642596e+00]])

In [68]:
pd.DataFrame(pca.fit_transform(car_scaler), columns=['PC1','PC2','PC3','PC4','PC5'])

PC1           PC2        PC3       PC4       PC5
0     37040.855194  -3144.048853  -2.767942  1.657015  1.322040
1     31203.261950  -3376.360144   2.219662 -2.375782 -0.762122
2     13222.107873  -4201.056303  -3.577684  4.774485 -0.436945
3     28079.543862  -3478.396717   0.382409  1.657190 -0.517506
4     54234.915742  -2212.253375   0.199814  2.540630 -0.463763
...            ...           ...        ...       ...       ...
2809 -48576.081713  18339.557973  13.200356 -0.490363 -1.247120
2810 -84674.876902  16593.456001   5.354746 -0.246929 -0.660051
2811 -39902.813509  21006.713468  10.778725  0.636291 -1.051091
2812 -39143.832405  22805.481121  14.170143  1.710612 -1.234752
2813   6614.025389  28923.360822  14.882293  0.731274 -1.186426

[2814 rows x 5 columns]

In [69]:
pca.explained_variance_ratio_ 

array([9.91505757e-01, 8.49423274e-03, 5.92162457e-09, 2.07538478e-09,
       5.45443785e-10])

- PCA(2)까지 의미있음

In [70]:
(pca.explained_variance_ratio_).sum()

0.9999999983854502

In [71]:
car_scaler.corr()

Price  Mileage(miles)  Registration_Year  \
Price                      1.000000       -0.459189           0.746840   
Mileage(miles)            -0.459189        1.000000          -0.390948   
Registration_Year          0.746840       -0.390948           1.000000   
Previous Owners           -0.188283        0.164313          -0.152570   
Engine                     0.096912        0.279453          -0.198822   
Doors                      0.020665       -0.032648           0.142310   
Seats                      0.012411        0.098564           0.036635   
Emission Class             0.668467       -0.320043           0.912551   
Fuel type_Diesel           0.154689        0.220828           0.155971   
Fuel type_Electric         0.039574       -0.026677           0.035873   
Fuel type_P_Hybrid         0.065864       -0.014607           0.035949   
Fuel type_P_plugin_Hybrid  0.107288       -0.000978           0.061555   
Fuel type_Petrol          -0.177778       -0.215663          -0.168726   
Body type_Convertible     -0.043681       -0.000188          -0.123666   
Body type_Coupe            0.035018        0.036671          -0.065308   
Body type_Estate          -0.006765        0.117170           0.022881   
Body type_Hatchback       -0.207715       -0.112706           0.012901   
Body type_MPV             -0.080113        0.053395          -0.045767   
Body type_Pickup           0.062722        0.017396           0.035586   
Body type_SUV              0.298315       -0.058445           0.191247   
Body type_Saloon           0.055329        0.106471          -0.107947   
Gearbox_Manual            -0.144507       -0.126003           0.162117   
Service history_Unknown   -0.075234        0.058394          -0.095221   
brand_ALFA                -0.041433        0.012520          -0.016412   
brand_AUDI                 0.070536        0.083445          -0.013802   
brand_BMW                  0.095339        0.098651          -0.055258   
brand_CHEVROLET           -0.034215       -0.007047           0.002341   
brand_CHRYSLER            -0.029587        0.018568          -0.039511   
brand_CITROEN             -0.029758       -0.024986           0.029668   
brand_DACIA                0.036121       -0.078701           0.085242   
brand_DAEWOO              -0.019791       -0.010103          -0.035222   
brand_DAIHATSU            -0.023473        0.016188          -0.029116   
brand_DODGE               -0.009496        0.019550          -0.014312   
brand_DS                   0.100333       -0.043128           0.070859   
brand_FIAT                -0.051070       -0.071213           0.012292   
brand_FORD                -0.067488        0.004252          -0.010880   
brand_HONDA               -0.088852        0.083198          -0.145259   
brand_HYUNDAI              0.011398       -0.045446           0.053809   
brand_INFINITI             0.039852       -0.014239           0.025388   
brand_JAGUAR               0.083375       -0.001587           0.024785   
brand_JEEP                -0.015735        0.014551          -0.016290   
brand_KIA                  0.048586       -0.055694           0.075734   
brand_LAND                 0.147053        0.017917           0.002293   
brand_LEXUS               -0.015317        0.046519          -0.066728   
brand_MASERATI             0.063046       -0.014041          -0.007890   
brand_MAZDA                0.001756       -0.023694          -0.010359   
brand_MERCEDES-BENZ        0.093868        0.060797          -0.023090   
brand_MG                   0.063842       -0.077558           0.018428   
brand_MINI                 0.007041       -0.036983          -0.010666   
brand_MITSUBISHI          -0.031435        0.037193          -0.032746   
brand_NISSAN               0.040069       -0.007172           0.043544   
brand_PEUGEOT             -0.019088       -0.078064           0.031211   
brand_PORSCHE             -0.003406        0.027653          -0.026159   
brand_PROTON              -0.02575